Sources/Matrix.swift

```diff
@@ -0,0 +1,370 @@
+ */
+
+import Foundation
+
+public struct Matrix<T: Arithmetic> {
+    
+    private(set) var rows : Int
+    private(set) var columns : Int
+    
+    internal var data: [T]
+    
+    init(rows: Int, columns: Int, data: [T]) {
+        self.rows = rows
+        self.columns = columns
+        self.data = data
+    }
+    
+    init(rows: Int, columns: Int) {
+        self.rows = rows
+        self.columns = columns
+        self.data = Array.init(repeating: T.zero, count: rows * columns)
+    }
+    
+    /**
+     :name:    size
+     :description:    returns a tuple with number of rows and columns
+     - returns:    (Int, Int)
+     */
+    public var size : (row: Int, column: Int) {
+        return (rows,columns)
+    }
+    
+    subscript(row: Int, column: Int) -> T {
+        get {
+            precondition(validIndex(row: row, column: column), MatrixError.invalid.localizedDescription)
+            return self.data[(row * columns) + column]
+        }
+        set {
+            precondition(validIndex(row: row, column: column ), MatrixError.invalid.localizedDescription)
+            self.data[(row * columns) + column] = newValue
+        }
+    }
+    
+    /**
+     :name:    validIndex
+     :description:    asserts that the (row,col) index is valid
+     - returns:    Bool
+     */
+    private func validIndex(row: Int, column: Int) -> Bool {
+        return row >= 0 && row < rows && column >= 0 && column < columns
+    }
+    
+    /**
+     :name:    column
+     :description:    gets the column using the index
+     - returns:    [Arithmetic]
+     */
+    func row(index: Int) -> [T] {
+        precondition(validIndex(row: index, column: 0), MatrixError.invalid.localizedDescription)
+        let row = Array(data[(index * columns)..<(index * columns) + columns])
+        return row
+    }
+    
+    /**
+     :name:    column
+     :description:    gets the column using the index
+     - returns:    [Arithmetic]
+     */
+    func column(index: Int) -> [T] {
+        assert(validIndex(row: 0, column: index), MatrixError.invalid.localizedDescription)
+        let column = (0..<rows).map { (currentRow) -> T in
+            let currentColumnIndex = currentRow * columns + index
+            return data[currentColumnIndex]
+        }
+        return column
+    }
+    
+    /**
+     :name:    transpose
+     :description:    mutates self to become the transpose matrix
+     - returns:    Void
+     */
+    mutating func transpose() {
+        var trans = Matrix.init(rows: self.columns, columns: self.rows, data: self.data)
+        for row in 0..<trans.rows {
+            for col in 0..<trans.columns {
+                trans[row,col] = self[col, row]
+            }
+        }
+        self = trans
+    }
+    
+    /**
+     :name:    transpose_copy
+     :description:    creates a copy of transpose matrix
+     - returns:    Matrix<T: Numeric>
+     */
+    func transpose_copy() -> Matrix<T> {
```

Comment:
The name should be `transposed()`. Swift follows this rule. For mutating functions they use infinitive verb like `sort()`, `reverse()`. For functions that return copy after performing an operation they use past tense form of the verb like `sorted()`, `reversed()`, `shuffled()`

Reply:
I agree with the need to follow swift naming conventions. Just to clarify, did you mean that `transpose_copy` to be `transposed`?

Reply:
Yeah, `func transposed() -> Matrix<T>`

Reply:
makes sense